File: otask-bot/bot/store.py
"""Простое файловое хранилище (JSON).

Одноаккаунтный режим: ключ otask берётся из конфига, поэтому здесь храним
только список чатов-подписчиков, флаг напоминаний и дедупликацию напоминаний.
"""
from __future__ import annotations

import asyncio
import json
import os
from typing import Any

from .config import config

_FILE = os.path.join(config.data_dir, "store.json")
_lock = asyncio.Lock()
_db: dict[str, Any] = {"chats": {}}
_loaded = False


def _blank_chat() -> dict[str, Any]:
    return {"reminders_enabled": True, "reminded_deadline": [], "reminded_overdue": []}


async def _ensure_loaded() -> None:
    global _loaded, _db
    if _loaded:
        return
    try:
        with open(_FILE, encoding="utf-8") as f:
            _db = json.load(f)
        if "chats" not in _db:
            _db["chats"] = {}
    except (FileNotFoundError, json.JSONDecodeError):
        _db = {"chats": {}}
    _loaded = True


async def _persist() -> None:
    os.makedirs(os.path.dirname(_FILE) or ".", exist_ok=True)
    tmp = _FILE + ".tmp"
    with open(tmp, "w", encoding="utf-8") as f:
        json.dump(_db, f, ensure_ascii=False, indent=2)
    os.replace(tmp, _FILE)


async def _chat(chat_id: int) -> dict[str, Any]:
    await _ensure_loaded()
    key = str(chat_id)
    if key not in _db["chats"]:
        _db["chats"][key] = _blank_chat()
    return _db["chats"][key]


async def register_chat(chat_id: int) -> None:
    """Запомнить чат как подписчика (после /start)."""
    async with _lock:
        await _chat(chat_id)
        await _persist()


async def set_reminders(chat_id: int, enabled: bool) -> None:
    async with _lock:
        chat = await _chat(chat_id)
        chat["reminders_enabled"] = enabled
        await _persist()


async def chats_for_reminders() -> list[int]:
    async with _lock:
        await _ensure_loaded()
        return [int(k) for k, c in _db["chats"].items() if c.get("reminders_enabled")]


async def mark_reminded(chat_id: int, task_id: str, kind: str) -> bool:
    """Пометить напоминание отправленным. True, если ещё не слали."""
    field = "reminded_deadline" if kind == "deadline" else "reminded_overdue"
    async with _lock:
        chat = await _chat(chat_id)
        lst: list[str] = chat[field]
        if task_id in lst:
            return False
        lst.append(task_id)
        if len(lst) > 1000:
            del lst[: len(lst) - 1000]
        await _persist()
        return True


async def clear_reminded(chat_id: int, task_id: str, kind: str) -> None:
    field = "reminded_deadline" if kind == "deadline" else "reminded_overdue"
    async with _lock:
        chat = await _chat(chat_id)
        lst: list[str] = chat[field]
        if task_id in lst:
            lst.remove(task_id)
            await _persist()

File: otask-bot/src/store.ts
import { mkdir, readFile, writeFile } from 'node:fs/promises';
import { dirname, join } from 'node:path';
import { config } from './config.js';

/**
 * Простое файловое хранилище (JSON). Достаточно для одного инстанса бота.
 * Для прод-нагрузки замените на БД, сохранив интерфейс.
 */
interface UserRecord {
  /** API-ключ otask, привязанный пользователем через /link. */
  apiKey?: string;
  /** Получать ли напоминания о дедлайнах. */
  remindersEnabled: boolean;
  /** id задач, по которым напоминание о дедлайне уже отправлено. */
  remindedDeadline: string[];
  /** id задач, по которым уже сообщили о просрочке. */
  remindedOverdue: string[];
}

interface DbShape {
  users: Record<string, UserRecord>;
}

const filePath = join(config.dataDir, 'store.json');
let db: DbShape = { users: {} };
let loaded = false;
let writeQueue: Promise<void> = Promise.resolve();

async function ensureLoaded(): Promise<void> {
  if (loaded) return;
  try {
    const raw = await readFile(filePath, 'utf8');
    db = JSON.parse(raw) as DbShape;
    if (!db.users) db.users = {};
  } catch {
    db = { users: {} };
  }
  loaded = true;
}

function persist(): Promise<void> {
  // Сериализуем записи, чтобы не было гонок при параллельных обновлениях.
  writeQueue = writeQueue.then(async () => {
    await mkdir(dirname(filePath), { recursive: true });
    await writeFile(filePath, JSON.stringify(db, null, 2), 'utf8');
  });
  return writeQueue;
}

function blankUser(): UserRecord {
  return { remindersEnabled: true, remindedDeadline: [], remindedOverdue: [] };
}

async function getOrCreate(chatId: number): Promise<UserRecord> {
  await ensureLoaded();
  const key = String(chatId);
  if (!db.users[key]) db.users[key] = blankUser();
  return db.users[key];
}

export const store = {
  /** Эффективный API-ключ: личный ключ пользователя или общий из конфига. */
  async getApiKey(chatId: number): Promise<string | undefined> {
    const u = await getOrCreate(chatId);
    return u.apiKey || config.defaultApiKey || undefined;
  },

  async hasOwnKey(chatId: number): Promise<boolean> {
    const u = await getOrCreate(chatId);
    return Boolean(u.apiKey);
  },

  async setApiKey(chatId: number, apiKey: string): Promise<void> {
    const u = await getOrCreate(chatId);
    u.apiKey = apiKey;
    await persist();
  },

  async clearApiKey(chatId: number): Promise<void> {
    const u = await getOrCreate(chatId);
    delete u.apiKey;
    await persist();
  },

  async setReminders(chatId: number, enabled: boolean): Promise<void> {
    const u = await getOrCreate(chatId);
    u.remindersEnabled = enabled;
    await persist();
  },

  /** Все чаты, у которых включены напоминания и есть доступ к API. */
  async chatsForReminders(): Promise<number[]> {
    await ensureLoaded();
    const out: number[] = [];
    for (const [key, u] of Object.entries(db.users)) {
      if (u.remindersEnabled && (u.apiKey || config.defaultApiKey)) out.push(Number(key));
    }
    return out;
  },

  /** Помечаем напоминание отправленным; возвращает true, если это новое (ещё не слали). */
  async markReminded(chatId: number, taskId: string, kind: 'deadline' | 'overdue'): Promise<boolean> {
    const u = await getOrCreate(chatId);
    const list = kind === 'deadline' ? u.remindedDeadline : u.remindedOverdue;
    if (list.includes(taskId)) return false;
    list.push(taskId);
    // Не даём спискам расти бесконечно.
    if (list.length > 1000) list.splice(0, list.length - 1000);
    await persist();
    return true;
  },

  /** Сброс отметки (например, когда дедлайн задачи изменился и снова в будущем). */
  async clearReminded(chatId: number, taskId: string, kind: 'deadline' | 'overdue'): Promise<void> {
    const u = await getOrCreate(chatId);
    const list = kind === 'deadline' ? u.remindedDeadline : u.remindedOverdue;
    const i = list.indexOf(taskId);
    if (i >= 0) {
      list.splice(i, 1);
      await persist();
    }
  },
};
